artifact/version/version.cpp

```diff
@@ -0,0 +1,124 @@
+
+#include <string>
+#include <vector>
+
+#include <common/common.hpp>
+#include <common/expected.hpp>
+#include <common/error.hpp>
+#include <common/json.hpp>
+
+#include <artifact/version/version.hpp>
+
+namespace artifact {
+namespace v3 {
+namespace version {
+
+namespace io = mender::common::io;
+namespace expected = mender::common::expected;
+namespace error = mender::common::error;
+namespace json = mender::common::json;
+
+
+const ErrorCategoryClass ErrorCategory {};
+
+const char *ErrorCategoryClass::name() const noexcept {
+	return "VersionParserErrorCategory";
+}
+
+string ErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	ParseError:
+		return "Parse error";
+	VersionError:
+		return "Wrong Artifact version";
+	FormatError:
+		return "Wrong Artifact format";
+	default:
+		return "Unknown";
+	}
+}
+
+error::Error MakeError(ErrorCode code, const string &msg) {
+	return error::Error(error_condition(code, ErrorCategory), msg);
+}
+
+
+expected::Expected<Version, error::Error> Parse(io::Reader &reader) {
+	// Read in all the bytes into a tmp buffer
+	std::vector<uint8_t> buf(1024);
+	io::ByteWriter bw {buf};
+
+	auto err = io::Copy(bw, reader);
+	if (err) {
+		return err;
+	}
+
+	auto expected_json = json::LoadFromString(mender::common::StringFromByteVector(buf));
+
+	if (!expected_json) {
+		return MakeError(ParseError, expected_json.error().message);
+	}
+
+	const json::ExpectedJson expected_version = expected_json.value();
+
+	if (!expected_version) {
+		return MakeError(ParseError, expected_json.error().message);
+	}
+
+	auto expected_obj = expected_version.value();
+
+	if (!expected_obj.IsObject()) {
+		return MakeError(
+			ParseError,
+			"Failed to unmarshal the version from the version JSON data. The received data does not contain a single top-level object");
+	}
+
+	auto version = expected_obj.Get("version");
+	if (!version) {
+		return MakeError(
+			ParseError,
+			"Failed to unmarshal the version from the version JSON data. No 'version' key found");
+	}
+	auto version_unwrapped = version.value().GetInt();
+	int version_;
+	if (!version_unwrapped) {
+		return MakeError(
+			VersionError,
+			"Failed to unmarshal the version from the version JSON data. The 'version' field needs to contain an integer");
+	}
+	version_ = version_unwrapped.value();
+
+	auto format = expected_obj.Get("format");
+	if (!format) {
+		return MakeError(
+			ParseError,
+			"Failed to unmarshal the version from the version JSON data. No 'format' field found");
+	}
+	auto format_unwrapped = format.value().GetString();
+	if (!format_unwrapped) {
+		return MakeError(
+			ParseError,
+			"Failed to unmarshal the version from the version JSON data. The 'format' fields needs to contain a string");
+	}
```

Comment:
Suggestion: You could shorten this by doing:

```c++
auto format = expected_obj.Get("format");
if (!format || !format.value().GetString()) {
	return MakeError(
		ParseError,
		"Failed to unmarshal the version from the version JSON data. 'format' field not found or not a string");
}
auto format_unwrapped = format.value().GetString();
```

And similar with `version` above. Not saying it's better though, verbosity is always good. Just a suggestion.